dora/core/common/src/main/java/alluxio/membership/DefaultEtcdMembershipManager.java

```diff
@@ -0,0 +1,154 @@
+   *
+   * @param conf              Alluxio configuration
+   * @param alluxioEtcdClient etcd client
+   */
+  public DefaultEtcdMembershipManager(AlluxioConfiguration conf,
+      AlluxioEtcdClient alluxioEtcdClient) {
+    mAlluxioEtcdClient = alluxioEtcdClient;
+  }
+
+  @Override
+  public void join(WorkerInfo workerInfo) throws IOException {
+    LOG.info("Try joining on etcd for worker:{} ", workerInfo);
+    WorkerServiceEntity entity =
+        new WorkerServiceEntity(workerInfo.getIdentity(), workerInfo.getAddress());
+    // 2) start heartbeat
+    mAlluxioEtcdClient.mServiceDiscovery.registerAndStartSync(entity);
+    LOG.info("Joined on etcd for worker:{} ", workerInfo);
+  }
+
+  @Override
+  public WorkerClusterView getAllMembers() throws IOException {
+    return getLiveMembers();
+  }
+
+  @Override
+  public WorkerClusterView getLiveMembers() throws IOException {
+    Iterable<WorkerInfo> workerInfoIterable = parseWorkersFromEtcdKvPairs(
+        mAlluxioEtcdClient.mServiceDiscovery.getAllLiveServices())
+        .map(w -> new WorkerInfo()
+            .setIdentity(w.getIdentity())
+            .setAddress(w.getWorkerNetAddress())
+            .setState(WorkerState.LIVE))
+        ::iterator;
+    return new WorkerClusterView(workerInfoIterable);
+  }
+
+  @Override
+  public WorkerClusterView getFailedMembers() {
+    return new WorkerClusterView(Collections.emptyList());
+  }
+
+  private Stream<WorkerServiceEntity> parseWorkersFromEtcdKvPairs(List<KeyValue> workerKvs) {
+    return workerKvs
+        .stream()
+        .map(this::parseWorkerServiceEntity)
+        .filter(Optional::isPresent)
+        .map(Optional::get);
+  }
+
+  private Optional<WorkerServiceEntity> parseWorkerServiceEntity(KeyValue etcdKvPair) {
+    try {
+      WorkerServiceEntity entity = new WorkerServiceEntity();
+      entity.deserialize(etcdKvPair.getValue().getBytes());
+      return Optional.of(entity);
+    } catch (JsonParseException ex) {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  @VisibleForTesting
+  public String showAllMembers() {
+    try {
+      WorkerClusterView registeredWorkers = getAllMembers();
+      WorkerClusterView liveWorkers = getLiveMembers();
+      String printFormat = "%s\t%s\t%s%n";
+      StringBuilder sb = new StringBuilder(
+          String.format(printFormat, "WorkerId", "Address", "Status"));
+      for (WorkerInfo entity : registeredWorkers) {
+        String entryLine = String.format(printFormat,
+            entity.getIdentity(),
+            entity.getAddress().getHost() + ":"
+                + entity.getAddress().getRpcPort(),
+            liveWorkers.getWorkerById(entity.getIdentity()).isPresent() ? "ONLINE" : "OFFLINE");
+        sb.append(entryLine);
+      }
+      return sb.toString();
+    } catch (IOException ex) {
+      return String.format("Exception happened:%s", ex.getMessage());
+    }
+  }
+
+  @Override
```

Comment:
Add @VisibleForTesting
can u help add the annotation onto etcdmembershipmanager as well? this should be used for testing only